sklearn/feature_selection/_from_model.py

```diff
@@ -247,18 +262,32 @@ def fit(self, X, y=None, **fit_params):
             Fitted estimator.
         """
         if self.max_features is not None:
-            if not isinstance(self.max_features, numbers.Integral):
-                raise TypeError(
-                    "'max_features' should be an integer between"
-                    " 0 and {} features. Got {!r} instead.".format(
-                        X.shape[1], self.max_features
+            if isinstance(self.max_features, numbers.Integral):
+                if self.max_features < 0 or self.max_features > X.shape[1]:
+                    raise ValueError(
+                        "When using an integral value, 'max_features' should be"
+                        f" between 0 and {X.shape[1]}. Got {self.max_features} instead."
                     )
-                )
-            elif self.max_features < 0 or self.max_features > X.shape[1]:
-                raise ValueError(
-                    "'max_features' should be 0 and {} features.Got {} instead.".format(
-                        X.shape[1], self.max_features
+                self.max_features_ = self.max_features
+            elif callable(self.max_features):
+                try:
+                    self.max_features_ = self.max_features(X)
+                    if not isinstance(self.max_features_, numbers.Integral):
+                        raise ValueError(
+                            "When `max_features` is a callable, it must return an"
+                            " integral value (e.g. `int`)"
+                        )
+                except TypeError:
+                    raise TypeError(
+                        "When 'max_features' is a callable, it must take only 'X' as"
+                        " input -- you passed a function with signature"
+                        f" {inspect.signature(self.max_features)}"
```

Comment:
I'm okay with not performing this check, and call `self.max_features`. I think the normal Python error is good enough.